runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/grpcclient/GrpcGetWorkerMetadataStreamTest.java

```diff
@@ -0,0 +1,338 @@
+    TestGetWorkMetadataRequestObserver requestObserver =
+        new TestGetWorkMetadataRequestObserver(
+            WorkerMetadataResponse.newBuilder()
+                .setMetadataVersion(1)
+                .addAllDirectPathEndpoints(DIRECT_PATH_ENDPOINTS)
+                .putAllGlobalDataEndpoints(GLOBAL_DATA_ENDPOINTS)
+                .build());
+    GetWorkerMetadataTestStub testStub = new GetWorkerMetadataTestStub(requestObserver);
+    stream =
+        getWorkerMetadataTestStream(
+            testStub, new AtomicLong(), new TestWindmillEndpointsConsumer());
+
+    assertTrue(streamRegistry.contains(stream));
+    stream.close();
+    assertFalse(streamRegistry.contains(stream));
+  }
+
+  @Test
+  public void testSendHealthCheck() {
+    TestGetWorkMetadataRequestObserver requestObserver =
+        Mockito.spy(
+            new TestGetWorkMetadataRequestObserver(
+                WorkerMetadataResponse.newBuilder()
+                    .setMetadataVersion(1)
+                    .addAllDirectPathEndpoints(DIRECT_PATH_ENDPOINTS)
+                    .putAllGlobalDataEndpoints(GLOBAL_DATA_ENDPOINTS)
+                    .build()));
+    GetWorkerMetadataTestStub testStub = new GetWorkerMetadataTestStub(requestObserver);
+    stream =
+        getWorkerMetadataTestStream(
+            testStub, new AtomicLong(), new TestWindmillEndpointsConsumer());
+    stream.sendHealthCheck();
+
+    verify(requestObserver).onNext(WorkerMetadataRequest.getDefaultInstance());
+  }
+
+  private static class GetWorkerMetadataTestStub
+      extends CloudWindmillServiceV1Alpha1Grpc.CloudWindmillServiceV1Alpha1ImplBase {
+    private final TestGetWorkMetadataRequestObserver requestObserver;
+    private @Nullable StreamObserver<WorkerMetadataResponse> responseObserver;
+
+    private GetWorkerMetadataTestStub(TestGetWorkMetadataRequestObserver requestObserver) {
+      this.requestObserver = requestObserver;
+    }
+
+    @Override
+    public StreamObserver<WorkerMetadataRequest> getWorkerMetadataStream(
+        StreamObserver<WorkerMetadataResponse> responseObserver) {
+      if (this.responseObserver == null) {
+        this.responseObserver = responseObserver;
+        requestObserver.responseObserver = this.responseObserver;
+      }
+
+      return requestObserver;
+    }
+
+    private void injectWorkerMetadata(WorkerMetadataResponse response) {
+      if (responseObserver != null) {
+        responseObserver.onNext(response);
+      }
+    }
+  }
+
+  private static class TestGetWorkMetadataRequestObserver
+      implements StreamObserver<WorkerMetadataRequest> {
+    private final WorkerMetadataResponse response;
+    private @Nullable StreamObserver<WorkerMetadataResponse> responseObserver;
+
+    private TestGetWorkMetadataRequestObserver(WorkerMetadataResponse response) {
+      this.response = response;
+    }
+
+    @Override
+    public void onNext(WorkerMetadataRequest workerMetadataRequest) {
+      responseObserver.onNext(response);
```

Comment:
what about removing this way to send responses and just having tests explicitly call injectWorkerMetadata?
I think it will be easier to see what tests are doing

Reply:
done.